From 611d94e7aadb1e04877d68fb4452a0d8f6236046 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Wed, 7 May 2025 15:23:47 +0200
Subject: [PATCH 834/954] fix(st-uart): aarch64: wait for UART ISR register TC
 bit for console flush
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Loop over the UART_ISR_TC bit to make sure the FIFO is emptied and all
the data is out of the shift register.
Fix the return value that should be void.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b0f022b7e06b74f4983cf266382ad48d6a3daf99 fix(st-uart): aarch64: wait for UART ISR register TC bit for console flush]
Change-Id: I3f3b125e110d6d252df5555209244d877499e0a6
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/456479
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/st/uart/aarch64/stm32_console.S | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/drivers/st/uart/aarch64/stm32_console.S b/drivers/st/uart/aarch64/stm32_console.S
index fc97f6d748..b28bb70994 100644
--- a/drivers/st/uart/aarch64/stm32_console.S
+++ b/drivers/st/uart/aarch64/stm32_console.S
@@ -213,7 +213,7 @@ endfunc console_stm32_core_getc
 	 * Function to force a write of all buffered
 	 * data that hasn't been output.
 	 * In : x0 - console base address
-	 * Out : return -1 on error else return 0.
+	 * Out : void.
 	 * Clobber list : x0, x1
 	 * ---------------------------------------------
 	 */
@@ -222,12 +222,11 @@ func console_stm32_core_flush
 	cmp	x0, #0
 	ASM_ASSERT(ne)
 #endif /* ENABLE_ASSERTIONS */
-	/* Check Transmit Data Register Empty */
-txe_loop_3:
+	/* Check Transmission Complete flag  */
+tc_loop:
 	ldr	w1, [x0, #USART_ISR]
-	tst	w1, #USART_ISR_TXE
-	beq	txe_loop_3
-	mov	w0, #0
+	tst	w1, #USART_ISR_TC
+	beq	tc_loop
 	ret
 endfunc console_stm32_core_flush
 
-- 
2.53.0

